jax/_src/scipy/signal.py

```diff
@@ -492,3 +492,110 @@ def welch(x, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
+  flat_batchsize = np.prod(batch_shape, dtype=np.int64)
+  x = x.reshape((flat_batchsize, segment_len, nframe))
+  out_length = step_size * (nframe - 1) + segment_len
+
+  nstep_per_segment = 1 + (segment_len - 1) // step_size
+  padded_segment_len = nstep_per_segment * step_size
+  pad_x = jnp.pad(x, ((0, 0), (0, padded_segment_len - segment_len), (0, 0)))
+  y = jnp.zeros((flat_batchsize, out_length), dtype=x.dtype)
+  for i in range(nstep_per_segment):
+    shifted = pad_x[:, step_size * i:step_size * (i + 1), :]
+    shifted = shifted.transpose((0, 2, 1)).reshape((flat_batchsize, -1))
+
+    leftpad = step_size * i
+    rightpad = max(0, segment_len - step_size)
+    shifted = jnp.pad(shifted, ((0, 0), (leftpad, rightpad)))[:, :out_length]
+
+    y += shifted
+  return y.reshape(tuple(batch_shape) + (-1,))
+
+
+@_wraps(osp_signal.istft)
+def istft(Zxx, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
+          input_onesided=True, boundary=True, time_axis=-1, freq_axis=-2):
+  # Input validation
+  if Zxx.ndim < 2:
+    raise ValueError('Input stft must be at least 2d!')
+  freq_axis = canonicalize_axis(freq_axis, Zxx.ndim)
+  time_axis = canonicalize_axis(time_axis, Zxx.ndim)
+  if freq_axis == time_axis:
+    raise ValueError('Must specify differing time and frequency axes!')
+
+  Zxx = jnp.asarray(Zxx, dtype=jax.dtypes.canonicalize_dtype(
+      np.result_type(Zxx, np.complex64)))
+
+  n_default = (2 * (Zxx.shape[freq_axis] - 1) if input_onesided
+               else Zxx.shape[freq_axis])
+
+  nperseg = jax.core.concrete_or_error(int, nperseg or n_default,
+                                       "nperseg: segment length of STFT")
+  if nperseg < 1:
+    raise ValueError('nperseg must be a positive integer')
+
+  if nfft is None:
+    nfft = n_default
+    if input_onesided and nperseg == n_default + 1:
+      nfft += 1  # Odd nperseg, no FFT padding
+  else:
+    nfft = jax.core.concrete_or_error(int, nfft, "nfft of STFT")
+  if nfft < nperseg:
+    raise ValueError(
+        f'FFT length ({nfft}) must be longer than nperseg ({nperseg}).')
+
+  noverlap = jax.core.concrete_or_error(int, noverlap or nperseg // 2,
+                                        "noverlap of STFT")
+  if noverlap >= nperseg:
+    raise ValueError('noverlap must be less than nperseg.')
+  nstep = nperseg - noverlap
+
+  # Rearrange axes if necessary
+  if time_axis != Zxx.ndim - 1 or freq_axis != Zxx.ndim - 2:
+    outer_idxs = tuple(
+        idx for idx in range(Zxx.ndim) if idx not in {time_axis, freq_axis})
+    Zxx = jnp.transpose(Zxx, outer_idxs + (freq_axis, time_axis))
+
+  # Perform IFFT
+  ifunc = jax.numpy.fft.irfft if input_onesided else jax.numpy.fft.ifft
+  # xsubs: [..., T, N], N is the number of frames, T is the frame length.
+  xsubs = ifunc(Zxx, axis=-2, n=nfft)[..., :nperseg, :]
+
+  # Get window as array
+  if isinstance(window, (str, tuple)):
+    win = osp_signal.get_window(window, nperseg)
```

Comment:
I think we should call `win = jnp.asarray(win)` to avoid doing the operations below at compile time.

Reply:
Done